Add Item#canPutIntoItemInventory

malte0811 opened the pull request:
Backpacks, things like the ImmersiveEngineering crates and storage cells from AE2/Refined Storage (I'll say "backpacks" in the rest of this description to refer to all such inventories) are known for causing issues when they are put inside each other. They are usually protected from being put into backpacks from the same mod (and from having filled shulker boxes put into them), but as soon as two mods adding backpacks without explicitly adding compatibility for each other are used together, NBT size issues are possible again. The vanilla Shulker box currently does not support blocking modded items at all.

This PR adds an event intended to prevent such issues: If a mod has items that should not be placed in a backpack it can subscribe to the event and set the result of the event to `DENY` to prevent those items from entering the backpack. To check whether an `ItemStack` is permitted in a backpack a mod can simply call `ForgeHooks.onStoreInItem` with the appropriate arguments. This way all mods adding backpacks would implicitly add compatibility for each other.

I added an `ItemStack` containing a rough representation of the container item and a list of the full contents of the inventory (before adding the new stack) to the event parameters since I felt that someone might need them for something. If you think that they are not necessary I will remove them.

Shulker boxes are prevented from being put into backpacks by the `ForgeInternalHandler`. I changed the vanilla behavior to allow shulker boxes without an inventory tag since they can't contain much NBT data. If such a change of vanilla behavior does not belong into Forge I will remove it and probably put it into a seperate mod. The ability to force allowing items into backpacks (like empty Shulker boxes) is the reason I used the event result rather than cancelling: An event handler with high or highest priority can set the result to `ALLOW`, thus overriding any later event handlers.

malte0811 committed:
Added the StoreInItemEvent
Called when something tries to place an item in an inventory that can retain its contents when picked up
Used to prevent NBT overflows by placing a bag in a different bag in a shulker box etc
The test mod prevents sticks from being placed in such an inventory unless it already contains blaze rods

malte0811 committed:
Changed the Shulker box blocker to allow empty Shulker boxes since those don't contain arbitrary NBT

malte0811 committed:
Re-added some imports I accidentally removed in the first commit

diesieben07 commented:
I am really not sure if I like this. The concept of an item having an inventory is already handled in an compatible way using capabilities. If you don't want items with inventories to be put in your inventory you can just check if the `ItemStack` provides the `IItemHandler` capability.

raoulvdberge commented:
Is your only argument for this NBT sizes? If so, that argument is invalid because one can use `Item#getShareTag`.

liach commented:
The idea for the event if good, but the way you implement it is pretty bad. You should add event before that method is called instead of patching that method itself.

liach commented:
The pull request is totally wrong, this event is only fired for shulker boxes!

kashike commented on src/main/java/net/minecraftforge/common/ForgeInternalHandler.java:
```diff
@@ -78,4 +82,18 @@ public void onDimensionUnload(WorldEvent.Unload event)
+        if (event.getResult()==Event.Result.DEFAULT && block instanceof BlockShulkerBox)
+        {
+            NBTTagCompound nbt = event.getInput().getTagCompound();
+            if (nbt!=null&&nbt.hasKey("BlockEntityTag")&&nbt.getCompoundTag("BlockEntityTag").hasKey("Items"))
```
Weird formatting.

kashike commented on src/test/java/net/minecraftforge/debug/StoreInItemEventTest.java:
```diff
@@ -0,0 +1,41 @@
+        if (event.getInput().getItem()==Items.STICK)
+        {
+            boolean deny = true;
+            for (int i = 0;i<event.getContents().size();i++) {
```
Weird formatting.

malte0811 replied:
Sorry, I'll fix all three. It is the formatting I use for my own projects, so I kind of automatically use it.

kashike commented on src/test/java/net/minecraftforge/debug/StoreInItemEventTest.java:
```diff
@@ -0,0 +1,41 @@
+        {
+            boolean deny = true;
+            for (int i = 0;i<event.getContents().size();i++) {
+                if (event.getContents().get(i).getItem()==Items.BLAZE_ROD)
```
Weird formatting.

malte0811 commented:
@diesieben07 Sorry, I didn't know that was the intended way to handle this. I used IE's toolbox as a reference, it uses a custom handler rather than capabilities (Probably left over from before capabilities existed). But from what I can see neither RefinedStorage's disks nor shulker box items have that capability, so the check you proposed wouldn't work. Another problem is that I don't want to block all items storing items, only those storing arbitrary items: The IE drill, for example, can store items, a drill head and some upgrades, but those generally have little to no NBT data.
@raoulvdberge Generally yes. From what I can see the same method is used to deserialize NBT from the network as is used for disk deserialization. Too large NBT compounds can cause bugs like BluSunrize/ImmersiveEngineering#1227, where exceptions where thrown when deserializing too large NBT compounds. If that can happen for serialization from/to disk, `getShareCompound` wouldn't help at all (I'll test this in a minute by having a TE generate a huge compound).
@liach I am not sure where you want me to put the call. If I put it in `Container` or `GuiContainer` the call would affect all inventories. I am not sure whether we would want to block shulker boxes from going into chests since that would be an even larger change of vanilla behavior. Every mod would have to put the call in their own `isValidForSlot` method if the inventory can be picked up. I don't think there are any other inventories in vanilla that can be picked up?

malte0811 commented:
I just checked, @raoulvdberge is right. NBT size by itself isn't an issue as long as the share compound is used. The only crash I was able to produce was due to NBT depth>512. I can't imagine someone reaching that limit except when intentionally trying to cause really nasty server crashes (the singleplayer world in which I tested this isn't even showing up in the menu any more). Is that a relevant "use"case for this PR? If it isn't and nobody can think of a better use case I will close this.

diesieben07 commented:
If someone does not expose the IItemHandler capability on their item then I would vote that is a bug in that mod.
As for allowing only specific items, there is currently some efforts into improving the IItemHandler API (https://github.com/MinecraftForge/MinecraftForge/issues/3344), I think part of that overhaul is allowing to detect which item can go into an inventory.

malte0811 commented:
This may be a stupid question, but are items like the shulker boxes supposed to have IItemHandlers (as an item in the inventory, where the items are not accessible)? Same goes for storage disks, AFAIK they aren't supposed to interact with generic item IO directly.
Only allowing specific items is possible with IE's implementation. The main issue isn't blocking the items but figuring out what items should be blocked (see my example of the drill above). I didn't read the whole discussion you linked since it is quite long, but I can't think of a way to differentiate items with arbitrary NBT size from ones without unless an explicite method for that is added.
Edit: I'll look into changing IE's items to use IItemHandler as soon as possible if Blu isn't opposed to using those.

malte0811 commented:
I just found another use case for this event, not a particularly big one though: IC2 uranium.
When it is in a players inventory the player is meant to get radiation poisoning, a very unpleasant potion effect. IC2 adds a containment box made from lead that prevents the radiation from harming the player. Currently it is possible to just take whatever is the cheapest backpack type item and safely transport highly radioactive materials in a woolen bag. IC2 could use the event to prevent radioactive materials from being put into any portable container other than their containment box. How systems like RefinedStorage would be handled would have to be decided by either the IC2 team (by adding some internal black/whitelist) or @raoulvdberge (by adding an high priority event handler).

raoulvdberge commented:
You want me to fire this event for *every* item inserted in the storage network? That will get slow REAL fast...

(I could only use this event when inserted into a Grid, and not the generic `INetwork#insertItem` method, but then you'd still be able to get the containment box or whatever in the system by other means)

malte0811 commented:
I have no idea how much IO a typical RefinedStorage system has, RS wasn't what I thought of when writing the event.
But I don't think you would have to fire it for every item: I can't think of any case where a mod wouldn't allow increasing the stack size of a stack that is already in the inventory (which I assume to be the majority of the input of an RS system, cobblestone, wood, other stuff from farms, ...). I will add a line to the javadoc saying that that assumption is meant to be valid. So you would only have to fire the event when putting a new item onto a disk which probably wouldn't be quite as bad for performance?

liach commented:
That assumption on radiation is incorrect: Modders hate to post events like these.

malte0811 commented:
So what you are saying is that the event is useless because no one will ever use it?
BTW, they don't have to fire the event directly, they could call the new method in ForgeHooks (is ForgeHooks meant to be accessed by mods? If not, where should I put the method so that mods can use it without being considered to use forge internals?). But that probably isn't your point.

liach commented:
ForgeHooks is used by patches. Modders will not remember to use that.

liach commented:
Patch before that accessible check

malte0811 commented:
> ForgeHooks is used by patches. Modders will not remember to use that.

Where should I put the method so that modders will remember to use it (at least with a higher probability)?

 > Patch before that accessible check

Sorry if I am being dumb, but what accessible check?

liach commented:
Patch where `canInsertItem` and `isItemValid` is called may be better.

malte0811 committed:
Fixed formatting
Added a condition to StoreInItemEvent to allow using ItemStoreEvent with high-throughput system like Refined Storage

malte0811 commented:
1. I just pushed a commit fixing the formatting (hopefully) and adding the condition of "not changing at runtime" to the documentation. To make this work for RefinedStorage systems I had to make the `contents` variable essentially useless, should I remove it?
2. @liach That would cause the change to affect all inventories, portable or not. As I said earlier being unable to put shulker boxes into chests (or any other inventories) would be a pretty major change of vanilla behavior. `canInsertItem` is called by the hoppers `ISidedInventory` handling code and by the sided inventory wrapper. Moving the check to those two places would probably do nothing more than block the items from all vanilla inventories as mods aren't supposed to use `ISidedInventory` any more. `isItemValid` is used in more places, I am not sure whether patching them without preventing Shulker boxes from being put in the player inventory as well.

JamiesWhiteShirt commented:
Just proposing a simpler alternative: How about adding a method to Item that can be used to determine if an item is a container, then make the shulker box use it?

The default implementation of this proposed method would return true only for shulker boxes. Then redefine the behavior of shulker boxes so that it only accepts items that are non-containers. It would then be the modder's responsibility to both override and check this method for container items.

Alternatively, change the semantics so that the method is instead used to determine if it can be placed in an item container, not if the item is a container.

LexManos commented:
> If you don't want items with inventories to be put in your inventory you can just check if the ItemStack provides the IItemHandler capability.

? If the ShulkerBox doesn't have it, we could add it.

raoulvdberge commented:
Not every "inventory" is an actual IItemHandler, look at AE or RS inventories for example...

malte0811 commented:
And it works the other way around too, the drill from IE for example can accept a few items (even though it doesn't do `IItemhandler` yet, I am working on that), but the items that can be put into it are so limited that they shouldn't cause any sort of issues.
@JamiesWhiteShirt Ok, I feel stupid now. I don't know why I didn't think of that. The only disadvantage I can think of would be that overriding other mods wouldn't be possible, but that doesn't really have too many use cases.

malte0811 committed:
Removed the StoreInItemEvent, replaced it with a method in Item

malte0811 commented:
I have implemented @JamiesWhiteShirt's solution on a [different branch](https://github.com/malte0811/MinecraftForge/tree/tmp). I think that solution is cleaner and it is definitely faster, but it adds more patches. Opinions on this? BTW, does anyone else find diffs of patches confusing?

JamiesWhiteShirt commented:
In your alternative branch you can reduce the patchwork slightly by hardcoding the behavior of the shulker box into the default Item implementation of `canPutIntoItemInventory` using an `instanceof` check. Ugly, but works and is used other places in Forge.

malte0811 commented:
You're right again. Since this is in a large block of forge-added methods I guess I don't need to "export" the 3 or 4 lines (due to the empty check) to ForgeHooks, the same was done for the method above mine? I'm just not used to working with patches, I try to write somewhat nice code.

malte0811 committed:
Reduced patch size slightly
Fixed both types of test item stacking in some situations

malte0811 commented:
I am updating my other PR to 1.12 right now. A lot of people don't seem to like this PR, should I update it (the version on the tmp branch) now as well or wait until someone decides whether this feature will be accepted at all?

HenryLoenwind commented:
I like it (don't care about event or callback). Just a couple of comments:

I don't see a need to put the inventory contents into the event, copying them takes time. Pass an IItemHandler if you want to enable funky tricks (e.g. only one stack of cobble allowed). Forcing this to be read only makes no real sense to me either, if a inventory doesn't want to be manipulated they can either not provide an IItemHandler when firing the event or a read only one.

I'd like a second constructors with World+BlockPos+BlockState instead of ItemStack. I'd implement this on Ender IO's Inventory Storage, which is a block until it is itemifyed.

With j8 hitting, you could probably implement this as `interface IRecursionAware { boolean canBeInserted(.....) default { return true; }; } [...] class item implements IRecursionAware [...]"`

malte0811 commented:
I don't really like the event based method myself, I just haven't gotten around to merging the tmp branch (it is in my fork if you want to see the code) into this one. I removed the `contents` parameter when I changed from using an event to using a method in Item. I don't think there is any good use-case for the parameter, if anyone can think of one I will add it.
I can't think of a clean way to add a BlockPos version when using a method in Item: I could either add add a second method to Item that takes these parameters and defaults to the other one with something like pickblock as the container or add the parameters to the first method and have one of the two 'sets' of parameters always be null. Neither of those alternatives sounds like good code to me, do you have a nicer idea?

HenryLoenwind commented:
I see three ways:

* Have one method with all params and javadoc that either none, one or the other set is filled.
* Have two methods.
* Have a parameter object that can be one of 2 classes. Introduces a new Object, but would allow people to subclass them if they want to add mod-specific data.

```
interface IRecursionAware {
    boolean canBeInserted(IRecursionAware.IParameter targetInventory, ItemStack item) default { return true; };
    public interface IParameter {}
    public class ItemStackParameter implements IRecursionAware.IParameter {
        ItemStack getStack() {
    [...]
    }
    public class BlockPosParameter implements IRecursionAware.IParameter {
        IBlockAccess getWorld() {
    [...]
        BlockPos getPos() {
    [...]
    }
}
```

malte0811 commented:
The second one is the worst IMO, especially since I suspect a lot of people would just override one and ignore the other unless someone reports it as a bug. I am not an expert on clean code, but I'll probably use the third option unless someone from the forge team tells me not to. Adding a class just to use it as a parameter for one method doesn't feel quite right, but allowing modders to add new 'categories' of inventories could be useful.

malte0811 committed:
Changed the second parameter of canPutIntoItemInventory to a new type, PortableInventoryLocation
This can contain information about the inventory the item is being put into

malte0811 commented:
I've implemented the third way suggested by @HenryLoenwind, using a class for the parameter rather than an interface. The parameter can be completely free of any information (base type, both inventory accessors as null), but should usually contain at least some information. I added an `IInventory` field in addition to an `IItemHandler` field to make access from GUI's easier (unless those are supposed to use `IItemHandler` as well now, `Slot` at least still uses `IInventory`).

mezz commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -688,12 +688,30 @@
++     * the implementations in {@link net.minecraftforge.items.PortableInventoryLocation} <br>
++     * @return true if the item should be allowed in the inventory
++     */
++    public boolean canPutIntoItemInventory(ItemStack stack, net.minecraftforge.items.PortableInventoryLocation location)
```
I think this should be `canBePutIntoItemInventory`.

mezz commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -688,12 +688,30 @@
 +    }
 +
++    /**
++     * Checks whether an item can be put into a container that is able to retain its contents as an item
```
I think this should be
`Returns whether this item can be stored in an item container (backpacks, Shulker boxes, etc).`

mezz commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -95,10 +95,7 @@
-import net.minecraftforge.event.DifficultyChangeEvent;
-import net.minecraftforge.event.ForgeEventFactory;
-import net.minecraftforge.event.ServerChatEvent;
+import net.minecraftforge.event.*;
```
Please do not change these imports, they are unrelated to your changes.

mezz commented on src/main/java/net/minecraftforge/common/ForgeInternalHandler.java:
```diff
@@ -27,7 +27,6 @@
 import net.minecraftforge.common.util.FakePlayerFactory;
 import net.minecraftforge.event.entity.EntityJoinWorldEvent;
 import net.minecraftforge.event.world.WorldEvent;
-import net.minecraftforge.fml.common.FMLLog;
```
Please do not change these imports, they are unrelated to your changes.

mezz commented on src/main/java/net/minecraftforge/items/PortableInventoryLocation.java:
```diff
@@ -0,0 +1,106 @@
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class PortableInventoryLocation
```
I am not sure this concept is good.
To check if your item is going into a subclass of `PortableInventoryLocation`, you have to already know about it and check using `instanceof` and then cast it.
If more subclasses are added by mods, there is no way for other mods to know about it.

malte0811 replied:
The idea with this is that most people don't need to know anything about the inventory (like the shulker boxes or normal backpacks). The subclasses are only relevant in cases like IC2 uranium which IMO shouldn't go into any backpacks but the containment box. In such cases it is just as you stated, the parameter has to be `ìnstanceof`-checked and cast to the right type. I can't think of any system that would not require a "default route": Determining whether an item should go into a specific (unknown) type of modded inventory is something that can't be done automatically IMO.
If you have a better idea I'd be glad to hear it.

mezz replied:
For that case, the IC2 uranium could always return false, and the IC2 containment box could check for uranium and ignore the result.

mezz replied:
I think `PortableInventoryLocation` is not much more clear than passing `Object` as a parameter, you will have to check it carefully anyway. I'm also not sure it's necessary.

mezz commented:
Looks like a good idea overall but needs a bit of refinement still, please see my review comments.
This will also need to be updated for 1.12.

malte0811 commented:
I'll start working on this and #3967 later today. I will be away for three weeks starting Saturday, so if I don't get these PR's done by then I won't be able to work on them for that time.
